EZP-31083: Improvements to Persistence/Cache to reduce (unfreeable) memory usage on Redis (mainly due to our tag usage) (#2840)
ezsystems/ezpublish-kernel: eZ/Publish/Core/Persistence/Cache/Readme.md

> # CORE DEV DOC: Persistence\Cache

Reviewer: Probably the whole document needs to be reviewed by the @ezsystems/documentation-team.

Reviewer 2: The real question is: do we really want to have this here instead of in the doc? The PR description says `Doc needed | no`, so I'm not sure what's the role of this file.

Author: This is doc for engineering and contributors who needs to know how things work internally and avoid repeating mistakes in the future, not for partners.

_So this is not for doc.ezplatform.com at all, but I think @alongosz suggested it to be doc team approved anyway._

Reviewer 2: Maybe it should be included in internal developer guidelines / code standards, once they are codified?
(Anyway, will give the file a quick look)

Author: Code / Core doc should really be close to the code imo, but otherwise 👍 It's basically doc for maintainers & contributors specifically, internal and external.

SPI Persistence Cache is a layer aiming to cache calls to Persistence (backend database) deemed costly under load. It
tries to balance this need up against the complexity and additional [system overhead](#guidelines-for-core-development) imposed by caching.

## Design

Handlers using `AbstractHandler` internally consist of one cache layer:

- "Shared cache": A Symfony Cache based cache pool, supporting a range of cache adapters like filesystem, Redis, Memcached.

>    Note: Due to being shared by design, clusters need remote cache, thus multi lookups are strongly advice to reduce round trip latency.

Reviewer 2: ```suggestion
   Note: Due to being shared by design, clusters need remote cache, thus multi lookups are strongly advised to reduce round trip latency.
```

> Handlers using `AbstractInMemoryHandler` / `AbstractInMemoryPersistenceHandler` in addition adds a second cache layer in front of "Shared cache":

Reviewer 2: ```suggestion
Handlers using `AbstractInMemoryHandler` / `AbstractInMemoryPersistenceHandler` in addition add a second cache layer in front of "Shared cache":
```

- "InMemory cache": A burst cache in PHP aiming at covering the most heavily used parts of the Content model to reduce repeated lookups to remote cache system.
   Note: It's not shared but per request/process. To keep risk of race condition negligible, it has own milliseconds ttl & item limits.

There are abstract test classes for the respective abstract classes above, these are opinionated and enforce conventions to:
- Avoid too much logic in cache logic _(e.g. warm-up logic)_, which can be a source of bugs.

> - Avoids having to write error prone test cases for each and every method.

Reviewer 2: ```suggestion
- Avoids having to write error-prone test cases for every method.
```

> _This makes sure cache layer is far less complex to maintain and evolve then what was the case in 1.x._

Reviewer 2: ```suggestion
_This ensures the cache layer is far less complex to maintain and evolve than what was the case in 1.x._
```

### Tags

> List of content tags that can be somewhat safely semi "officially" used to clear related entities in cache:

Reviewer 2: ```suggestion
List of content tags that can be somewhat safely "semi-officially" used to clear related entities in cache:
```

- `content-<id>`: Cache tag which refers to Content/ContentInfo entity.
- `location-<id>`: Cache tag which refers to Locations and/or their assigned Content/ContentInfo entities.

> - `location-path-<id>`: Like tag above but applied to all Content/Locations in the subtree of this ID, so can be used by tree operations.

Reviewer 2: ```suggestion
- `location-path-<id>`: Like the tag above but applied to all Content/Locations in the subtree of this ID, so it can be used by tree operations.
```

> - `content-fields-type-<type-id>`: Cache tag which refers to entries containing fields data. It's used on ContentType changes that affect all content items of type.

Reviewer 2: ```suggestion
- `content-fields-type-<type-id>`: Cache tag which refers to entries containing Field data. It's used on Content Type changes that affect all Content items of the type.
```

_For further tags used for other internal use cases, see the *Handlers for how they are used._

## Guidelines for core development

> ### Shared Cache: When to use, and not to use

Reviewer 2: ```suggestion
### Shared Cache: When to use, when not to use
```

It's worth noting that shared cache comes at a cost:
- Increased complexity
- Latency per round trip
- Memory use

Because of that, *typically* avoid introducing cache if:
- Lookup is per user => _it will consume a lot of memory and have very low hit ratio_

> - For drafts => _usually belongs to a single user & is short lived_

Reviewer 2: ```suggestion
- For drafts => _usually belongs to a single user and is short-lived_
```

- Infrequently used lookups
- Lookups that are fast against DB even under load _(see also notes in "Possible future plans")_

> ### Tags: When to use, and not to use

Reviewer 2: ```suggestion
### Tags: When to use, when not to use
```

Like cache, tags also comes at a cost:
- Slower invalidation
- Memory cost
  - _E.g.: ATM on RedisTagAwareAdapter tag relation data is even non-expiring as it needs to guarantee surviving cache._

> For those reasons, only introduce use a tag if:

Reviewer 2: ```suggestion
For those reasons, only introduce use a tag:
```

- Mainly to represent an entity _(e.g. `content-<id>`)_
- Only if it's represented on many different cache keys or if a key can have a lot of different variants.
    - _Tip: Otherwise prefer to delete by cache key(s) when cache clear is needed, it will be faster and consume less memory._

### Possible future considerations

> Ideally, the cache should for best effect, be in place to cache results of complex calculations based on multiple backend

Reviewer 2: ```suggestion
Ideally and for best effect, the cache should be in place to cache results of complex calculations based on multiple backend
```

> lookups. Caching lightweight SPI lookups sometimes might not give benefits, while still consume Redis/Memcached memory.

Reviewer 2: ```suggestion
lookups. Caching lightweight SPI lookups sometimes might not provide benefits, while still consume Redis/Memcached memory.
```

> This is why it's been discussed to rather move some of the caching to API instead in a future major release.

Reviewer 2: ```suggestion
This is why it's been discussed to move some of the caching to API instead in a future major release.
```

> However, that would require that permissions are split into own Repo layer => In order to not end

Reviewer 2: ```suggestion
However, that would require that permissions are split into their own Repo layer => In order to not end
```
